### ISSUE TITLE
Use of Auth2 introspect as default

### ISSUE THREAD

**Issue by Martin-Rosner** [IHE/IT-Infrastructure#106]
The current draft relies heavily on the use of JWT tokens (which are mandatory) as OAuth2 access tokens. The limiting factor is that all authorization decisions are to be made when the JWT authorization token is handed out, as these are codified inside the token. This means that the token contains information who the client is and what it may do. It does not take into account what the client may do at a specific server. Say for example there are two servers, named A and B, and both implement API X and Y. The token cannot indicate that a client may invoke API X at server A and API Y on server B. 

To enable this, the proposal is to rely as a default on Auth2 introspect (RFC7662). The introspect call is authenticated, such that the authorization server knows which specific client-server combination to authorize. Moreover, with introspect, the access token is opaque to the clients. The structure is to be defined and understandable only by the authorization server. This gives implementation freedom that server may still use JWT, but it may also hand-out large random numbers.
Using introspect makes that the IUA could fall back fully on widely supported OAuth2 mechanisms, rather than a few IETF drafts (that are not RFCs yet) : OAuth2 authorization framework; OAuth2 bearer token headers; OAuth2 introspect
 
The SAML and JWT tokens are still possible as options, using the custom authorization header methods. Clients that do not support these types of tokens would still be able to treat them as opaque bearer tokens and offer them to the introspect endpoint to get them validated and receive the contents (in JSON format). 
The introspect result can be identical to the IUA JWT contents in structure, including the IUA and BPPC extensions, with the addition of the active field as defined in RFC7662.

**Comment by msmock** [comment]
Using introspection in IUA would be a breaking change to the current profile. I personally like it, since the access token is just an ID and the resource server exchanges the access token to the data it requires to authorize. But this would really be a big change and would also not be compatible with SMART on FHIR. Thus I didn't follow this path, when reworking the IUA profile.

**Comment by JohnMoehrke** [comment]
breaking changes are allowed during Trial Implementation. Now is a good time to make a breaking change that is the right thing to do. You can alternatively define an option that uses introspection.

**Comment by Martin-Rosner** [comment]
It is recommended to use the introspect as a default for this profile for the reasons stated above. As such, indeed this is a breaking change. Sounds like we are not averse to breaking changes at this stage of development. How do we progress to affect this recommendation?

**Comment by joostreuzel** [comment]
SMART On FHIR mandates the use of bearer tokens (see http://hl7.org/fhir/smart-app-launch/index.html#step-3-app-exchanges-authorization-code-for-access-token). 

In the current IUA specification bearer tokens are an option, so to be compatible with SMART, implementers must implement that option. In that sense I would argue that by default the current IUA specification is not compatible with SMART, because of the difference in token types.

Moreover, the SMART specifications explicitly do not indicate the format of the access token. However, they also do not indicate how the access token is validated. OAuth2 Introspect would be a compatible means. Leaving it up to the authorization server to choose its token structure.

As such, my suggestion would be to promote the bearer token and introspect API to the default way of authorizing. The IHE-JWT and IHE-SAML token format would be modelled as options. Interesting about this approach is that if a client would not understand the IHE-JWT or IHE-SAML format, or would not have the logic to validate them correctly, it can still fallback to the introspect API to get these tokens validated.

The change is indeed breaking from a specification point of view, because clients and servers must understand the bearer tokens. That said, existing implementations that rely on the IHE-JWT or IHE-SAML format would not be completely incompatible, as they would just utilize those options...

**Comment by JohnMoehrke** [comment]
Note that an important distinction IUA has vs SMART... is that IUA does not mandate a tight relationship between the authorization server and the resource server. SMART does. Thus the choice of token types can be something totally hidden from the client in SMART. In IUA this is not the case, it might be, but it is not mandated.

We did discuss today that we could add an option for introspection. This would be a signficnant add to the profile, which will be difficult to get done by our November deadline. But we will try. 

@joostreuzel can you help with the addition? I can do the IHE Profiling narrative structural stuff, but don't know enough about OAuth to write any details.  How much 'profiling' of the OAuth introspection are needed? Can we just point at an OAuth defined transaction and just say the result is our JWT definition?

**Comment by lcmaas** [comment]
Introspection becomes complicated when there is not a tight relationship between the AS and the resource server, since each resource server needs to be able to authenticate to each AS in order to access the protected introspection endpoint. RFC 7662 requires authentication of the resource server but doesn't require a specific authentication method. Unfortunately, you can't use IUA in its current form to authenticate to the introspection endpoint or it creates a circular introspection dependency. To work at scale, you'd need to automate the registration and authentication of actors in this community of AS's and RS's. BTW, extensive work has been done in this area.

**Comment by joostreuzel** [comment]
I would argue that with the JWT and SAML specification there is also relation between the authorization server and the resource server, as the resource server will need the right certificates (and trust in those) to validate the JWT and SAML tokens. These certificates are typically pulled (regularly) from a well-known endpoint from the trusted authorization service. In that sense, introspection does not really make difference: clients will know their trusted authorization services.

With introspect, the authorization server has knowledge of the resource servers as these will need to perform authenticated calls towards the introspection endpoint. With JWT or SAML that is not necessarily the case. Depending on the scenario this knowledge may actually be good thing: the authorization server may leverage that knowledge in its authorization logic and thus introspect response (as we do). It provides the opportunity to control which resource servers are part of your ecosystem. Resource services in such cases must be pre-registered. 

For cases where there is no explicit need for the authorization server to know the resource servers, there is RFC7591: dynamic client registration that allows resource servers to register themselves.

**Comment by joostreuzel** [comment]
@JohnMoehrke I would be happy to help out. RFC7662 does not leave a lot of room for profiling (as it is so simple). The elements we need to specify is the way of authentication to the endpoint (basic authentication is typically supported), and refer to RFC7591 for client registration. The latter I would provide as recommendation.

**Comment by msmock** [comment]
During yesterdays IUA bi-weekly meeting, we decided to add the introspection way as an option, but not as a mandatory requirement. Since we need to finish the supplement till end of October and I'm running out of time since end of October is also the due date for the changes to swiss national extensions to be finished. 

Who can provide text blocks specifying the introspection requirements and the new required transaction to add this option?

**Comment by JohnMoehrke** [comment]
I will provide the supplement structure, and let @joostreuzel fill in the details. We should make it as terse as possible.

**Comment by JohnMoehrke** [comment]
@joostreuzel please see that I have added the supplement framework for the new transaction. Please check all my text around Introspection. The text I put in there is based on my executive level understanding. Most of what you need to do is at the end of the file in the new transaction ITI-102. It is setup with the typical IHE supplement template as a request and response. I also did include the link to the JWT constraints we have documented in 71, so please don't replicate that detail, as we can use it by reference.

**Comment by JohnMoehrke** [comment]
Reminder my pull request was about adding the IHE profile supplement layout. I do not know enough about OAuth. So if the client is not involved, then please go ahead and make that change.

**Comment by msmock** [comment]
To be of use in Swiss EPR, the introspection should also be able to response a XUA compliant access token in SAML format. Can you take that into account?

**Comment by joostreuzel** [comment]
The introspect call is token format agnostic. You simply pass it the bearer token as is. The Authorization Server (which handed out the token in the first place) is expected to understand its own formats... That is actually one of the reasons of the introspect endpoint: it makes both authorization client as well as resource server agnostic to the format of the token. With introspect, the only actor that needs to understand the token format is the Authorization Server, which both creates and interprets the token structure.

**Comment by joostreuzel** [comment]
see #112 for the pull request

**Comment by msmock** [comment]
Your last commit did not update the table of content. Can you do that?

**Comment by JohnMoehrke** [comment]
In this new model of publishing with markdown... do we want to make the table of content more sparse? I think it is not needed in an html publication to cotiunue to have anything other than H1 table of content.

**Comment by joostreuzel** [comment]
I will do that...

That said. The text describes now clearly the three options for token format and validation (JWT, SAML, or via introspect). JWT is required, the others are not as agreed.

During writing, I started to be convinced that implementations should be provided with the freedom to choose one or more of these three options. The three options would become all optional, but there is an additional constraint that at least one of the three must be supported. Resource Servers that receive a token they cannot validate should simply refuse the request. Without this freedom, servers that would implement the SAML option would also (and often unnecessarily) need to implement JWT support.

Perhaps this is something to discuss on a next call.

**Comment by JohnMoehrke** [comment]
please create a new issue for the changing of options.  This issue should be closed with the addition of the introspection option.

**Comment by joostreuzel** [comment]
new issue created: #115

**Comment by joostreuzel** [comment]
toc has been updated in 50daa2c2a4077d04ed972c5d5bbf4e4d436cab3a  The comments before and after hint to the use of the markdown-toc extension to vscode...

**Comment by JohnMoehrke** [comment]
is this issue done?

**Comment by joostreuzel** [comment]
I think the initial question (making introspect default) is not what we resolved here. Renaming the issue to "add introspection option" would be a better match, and we could close this issue. 

In the next telco, I would still like to have the discussion on what is required and what is optional with respect to token formats/introspect. See #115.

**Comment by JohnMoehrke** [comment]
@joostreuzel will write up an "Open Issue" for inclusion in the public comment.
@msmock will assure the open issue has counter arguments against mandatory introspection in JWT environments.

**Comment by msmock** [comment]
@john: I would expect a strong justification for making introspection mandatory first, since Jost clearly stated that it is easy for phillips to support the JWT option, but they just don‘t want to do it.

**Comment by joostreuzel** [comment]
I would like to stress that the JWT option _is not easy_ for Philips. There is explicit reliance on reference tokens, because of three reasons:
* The tokens do not contain any data. Any party that gets hold of a JWT token has access to the (privacy and security) sensitive data contained in that token. This includes resource servers which access has been revoked, intermediate routers, etc. Our legal/regulatory experts have some serious issues with this.
* Our authorization models make use of constructs that are not encodable in the current JWT format. Think of per resource server scope settings (while the token may be used with several resource servers simultaneously).
* Resource Servers dealing with introspect will be able to handle all types of tokens (XUA, JWT, opaque). This allows them to be used in more different solution configurations without code changes.

While JWT may remove the need for introspect calls, it is not as powerful functionally, privacy and even security-wise. I would rather state that for cases where JWT is used, the introspect call is trivial to implement as it would basically just do the token validation, audience check and then return the JWT contents. Simple to implement and offer as a generic (micro-) service.

In short this not a matter of “just not willing to”. I would rather provide the counter argument: as it so trivial to implement introspect for JWT tokens, why not just offer it as a generic service in the Swiss project?

**Comment by msmock** [comment]
I can't see this as a strong justification:

- access token are encrypted. Thus it is not justified to claim that "Any party that gets hold of a JWT token has access to the (privacy and security) sensitive data contained in that token.” 

- I see. It’s not up to me to criticize your design decisions. But you are mixing up authorization and policy enforcement in the resource servers and break the foundations of the standard by adding extra functionality. It’s up to Phillips to do so, but I don’t see why we should manifest that for the IHE community. 

- Resource server would also by prepared for the 3 token types, if we we leave the 3 options, not making introspection mandatory. 


Introspection might be trivial to implement from a coding perspective. But that’s not an argument for making it mandatory in general, since it may have a big impact in real life projects. 

In the Swiss electronic patient dossier IUA is used in a cross community context. To minimize the required cross community transactions and the related trade off (operational, organizational, testing, etc.), we follow a representational state transfer approach, i.e., we provide all the data in the access token which are required by the resource server to enforce the access policies. This approach is not Swiss specific and quite common in real life projects. Currently we use XUA access token only, but we are preparing to extend the EPR for mobile access using a JWT token. Thus, the introspection option is not suitable for the Swiss electronic patient dossier. 

The ordinances of the law of the Swiss electronic patient dossier rely on the IHE profiles (currently 16 IHE ITI profiles). Vendors must implement the mandatory transactions of the underlying IHE profiles, as well as the Swiss national extensions. They must run conformance tests to proof compliance against a Gazelle environment (so called SIA tests). The tests cover the connectathon tests for the IHE profiles as well as the additional tests to verify compliance with the national extensions.

Making the introspection option mandatory requires the Swiss vendors to implement and proof compliance to the introspection options, although it is useless in the context of the Swiss electronic patient dossier context, since it would exchange a JWT to the same JWT or a XUA assertion to exactly the same XUA assertion. 

To summarize, the introspection option does not meet the requirements of the Swiss electronic patient dossier and if mandatory, causes a high trade off (operational, organizational, testing, etc.). 

If there are doubts on the interoperability of the current IUA.b requirements, I suggest to make the JWT option mandatory and all other options mandatory, as it is in the current IUA version.

**Comment by joostreuzel** [comment]
> access token are encrypted. Thus it is not justified to claim that "Any party that gets hold of a JWT token has access to the (privacy and security) sensitive data contained in that token.”

In the current IUA specification JWT token encryption is optional. We may want to add a statement in transaction ITI-72 that resources servers MUST be able to handle JWE tokens.

> I see. It’s not up to me to criticize your design decisions. But you are mixing up authorization and policy enforcement in the resource servers and break the foundations of the standard by adding extra functionality. 

As perhaps expected, I don't agree there. The Authorization Server is a policy decision point, not the policy enforcement point. That does not change in our designs with introspect. It is just that policy decisions can be communicated more precisely using introspect than encoded with JWT tokens, as policy determination can happen "just in time" in the authorization process and catered towards specific resource server/client combinations. Policy enforcement still takes place at the resource servers.

Note that I do understand your case for a RESTful transaction that is all-encompassing in the sense that it carries all the necessary elements to handle the request, including any supporting (authorization) data. This reduces dependencies between systems to a bare minimum. Let me stress that it is not my intention to break this possibility. JWT is a valuable option in many situations.

Our primary use-case is different. We are not constructing a nation-wide EPD. What we try to do is construct what we call "end-2-end solutions" from a multitude of systems with varying capabilities, ranging from questionnaire providers, EMRs, (out-of-hospital) patient monitoring systems, imaging modalities, IoT solutions, etc. The challenge is to have these integrate, typically tailored to the context of a single customer or health system. Flexibility is required as different combinations of those products will be necessary for different customers. HTTP-based RESTful APIs are necessary to deal with the increase of cloud-based systems. 

Many of these customers already have a user authentication mechanism in place that is able to hand out access tokens (e.g. based on OpenID). These tokens will differ in structure depending on the customer's installed base. What we offer is an authorization layer on top of the authentication systems that besides some complex authorization policies uses introspect to provide a harmonized view on these tokens. The resource servers, relying on introspect, do not need to understand the (customer specific) access token structures. Forcing the use of JWT tokens does not make sense in these situations as it would introduce a layer of complexity (e.g. token exchange mechanisms) that are unnecessary and error-prone.

Different use-cases call for different approaches. Where JWT may be more suitable in the cross-organization/community projects, introspect helps us deal with the variances in legacy systems. As such I feel this choice may need to be made at a project level, not at a general IUA level.

**Comment by joostreuzel** [comment]
@JohnMoehrke should I reopen issue 2 in the document, or shall I create a new issue for this? Moreover, should I refer to this and the other threads from the issue statement?

**Comment by msmock** [comment]
> Different use-cases call for different approaches. 
> 
That was my position from the beginning. So we agree upon not making introspection mandatory.

**Comment by JohnMoehrke** [comment]
@msmock correct. We are NOT making introspection mandatory at this time. We are just adding an open issue that explains our current situation of three modes all optional, while asking for feedback on if introspection should be made mandatory. So we are now trying to come up with the wording of the open issue..

@joostreuzel a new issue.

**Comment by joostreuzel** [comment]
> That was my position from the beginning. So we agree upon not making introspection mandatory.

I think the current proposal makes sense: make all mechanisms for claim derivation from tokens (JWT/SAML/Introspect) optional and move that choice to individual projects. Vendors that claim maximum conformance can always opt to implement all three.

**Comment by joostreuzel** [comment]
@JohnMoehrke OK, will update the pull request.
